donut/modules/newsgroups/helpers.py

```diff
@@ -0,0 +1,208 @@
+from donut.modules.groups import helpers as groups
+
+
+def get_past_messages(group_id, limit=5):
+    """Returns a list of past sent messages"""
+
+    query = """
+    SELECT newsgroup_post_id, subject, message, time_sent, user_id, post_as
+    FROM newsgroup_posts WHERE group_id=%s ORDER BY time_sent DESC LIMIT %s
+    """
+    with flask.g.pymysql_db.cursor() as cursor:
+        cursor.execute(query, (group_id, limit))
+        return cursor.fetchall()
+
+
+def get_newsgroups():
+    """Gets all newsgroups."""
+
+    query = '''
+    SELECT group_name, group_id FROM groups 
+    WHERE newsgroups=1
+    '''
+    with flask.g.pymysql_db.cursor() as cursor:
+        cursor.execute(query)
+        return cursor.fetchall()
+
+
+def get_my_newsgroups(user_id, send=False):
+    """Gets groups user is a member of (includes indirectly).
+    If send is true, only gets newsgroups user can send to
+    """
+
+    query = '''SELECT DISTINCT g.group_name, g.group_id
+    FROM positions p LEFT JOIN position_relations pr
+    ON p.pos_id=pr.pos_id_to
+    INNER JOIN position_holders ph ON ph.pos_id=p.pos_id OR
+    pr.pos_id_from=ph.pos_id
+    INNER JOIN groups g ON p.group_id=g.group_id
+    WHERE ph.user_id=%s AND g.newsgroups=1'''
+    if send:
+        query += ''' AND p.send=1 
+        UNION DISTINCT SELECT group_name, group_id
+        FROM groups 
+        WHERE groups.newsgroups=1
+        AND groups.anyone_can_send=1'''
+    with flask.g.pymysql_db.cursor() as cursor:
+        cursor.execute(query, user_id)
+        return cursor.fetchall()
+
+
+def get_user_actions(user_id, group_id):
+    """Gets allowed actions for user."""
+
+    if not user_id:
+        return None
+    query = """
+    SELECT p.send AS send, p.control AS control, p.receive AS receive
+    FROM positions p LEFT JOIN position_relations pr
+    ON p.pos_id=pr.pos_id_to
+    INNER JOIN position_holders ph ON ph.pos_id=p.pos_id OR
+    pr.pos_id_from=ph.pos_id
+    WHERE ph.user_id=%s AND p.group_id=%s"""
+    with flask.g.pymysql_db.cursor() as cursor:
+        cursor.execute(query, (user_id, group_id))
+        return cursor.fetchone()
+
+
+def apply_subscription(user_id, group_id):
+    '''
+    Inserts into groups table with the position "Applicant"
+    '''
+    query = """
+    INSERT INTO group_applications(group_id, user_id)
+    VALUES (%s, %s) ON DUPLICATE KEY update user_id=user_id
+    """
+    with flask.g.pymysql_db.cursor() as cursor:
+        cursor.execute(query, (group_id, user_id))
+
+
+def unsubscribe(user_id, group_id):
+    if not user_id:
+        return None
+    query = """
+    UPDATE position_holders 
+    SET receive = 0
+    WHERE user_id = %s AND pos_id IN (
+    SELECT p.pos_id FROM positions p LEFT JOIN position_relations pr
+    ON p.pos_id=pr.pos_id_to
+    INNER JOIN position_holders ph ON ph.pos_id=p.pos_id OR
+    pr.pos_id_from=ph.pos_id
+    WHERE ph.user_id=%s AND p.group_id=%s)
+    """
+    query += POSITIONS_QUERY + ')'
```

Comment:
What is `POSITIONS_QUERY`? I'm confused as to how this code even runs

Reply:
typo